### ISSUE TITLE
Memory leak when using kerberos

### ISSUE THREAD

**Issue by drorgu** [edenhill/librdkafka#1934]
Hi, 

Using librdkafka-0.11.3.

Using Linux: CentOS release 6.9 (Final)

These are the Kafka Parameters: 
[
{"Param":"bootstrap.servers","Value":"X.X.X.X:9092,Y.Y.Y.Y:9092,Z.Z.Z.Z:9092"},
{"Param":"batch.num.messages","Value":"50"},
{"Param":"queue.buffering.max.ms","Value":"200"},
{"Param":"acks","Value":"all"},
{"Param":"debug","Value":"security"}, 
{"Param":"sasl.mechanisms", "Value":"GSSAPI"},
{"Param":"compression.codec","Value":"lz4"},
{"Param" : "security.protocol","Value" : "SASL_PLAINTEXT"},
{"Param": "sasl.kerberos.principal", "Value": "X@Y.Z"}, 
{"Param":"sasl.kerberos.keytab","Value": "/XXX.keytab"},{"Param":"sasl.kerberos.service.name","Value":"kafka"}
]

For some reason there is a memory leak because of the kerberos parameters. When using all above parameters except for the last three there is no memory leak.

The message are sent and received in the kafka cluster.

Could this have been fixed in version 11.4? I see a memory leak fix for ticket #1534.

Looking at librdkafka debug, there are three kafka nodesabove. Two are authenticated by Kerberos (but there is a failed kinit). The third nodes is immediately gets "connection refused".

Here is the debug related log messages to the two kafka nodes that get authenticated by Kerberos: 

1533817048.307 RDKAFKA-7-AUTH: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Auth in state APIVERSION_QUERY (handshake supported)
1533817048.310 RDKAFKA-7-SASLMECHS: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Broker supported SASL mechanisms: GSSAPI
1533817048.310 RDKAFKA-7-AUTH: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Auth in state AUTH_HANDSHAKE (handshake supported)
1533817048.310 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Initializing SASL client: service name kafka, hostname X.X.X.X, mechanisms GSSAPI, provider Cyrus
1533817048.310 RDKAFKA-7-SASLREFRESH: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Refreshing SASL keys with command: kinit -S "kafka/X.X.X.X" -k -t "file.keytab" A@B.C
1533817048.693 RDKAFKA-3-SASLREFRESH: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: SASL key refresh failed: kinit -S "kafka/X.X.X.X" -k -t "file.keytab" A@B.C: exited with code 1
1533817048.694 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: My supported SASL mechanisms: CRAM-MD5 LOGIN DIGEST-MD5 ANONYMOUS PLAIN GSSAPI EXTERNAL
1533817048.777 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Send SASL frame to broker (655 bytes)
1533817048.781 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Received SASL frame from broker (112 bytes)
1533817048.784 RDKAFKA-7-LIBSASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: CB_CANON: flags 0x3, "A@B.C" @ "(null)": returning "A@B.C"
1533817048.784 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Send SASL frame to broker (0 bytes)
1533817048.787 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Received SASL frame from broker (36 bytes)
1533817048.787 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Send SASL frame to broker (32 bytes)
1533817048.787 RDKAFKA-7-SASL: rdkafka#producer-1: [thrd:sasl_plaintext://X.X.X.X:9092/bootstrap]: sasl_plaintext://X.X.X.X:9092/bootstrap: Authenticated as A@B.C using GSSAPI (gssapiv2)

No related debug log messages to the third node, but there is err_cb message: 

err_cb - error received (-195): reason = sasl_plaintext://Z.Z.Z.Z:9092/bootstrap: Connect to ipv4#Z.Z.Z.Z:9092 failed: Connection refused

Thanks,
Dror

**Comment by edenhill** [comment]
How are you measuring memory leakage?
How much memory is being leaked?

**Comment by drorgu** [comment]
Hi, 
First, I'd like to mention that I installed 11.5 and the issue still occurs.
I am measuring the memory leak with a bash script utilization top.
I'm still investigating this for a few days now and I did find a few issues on my side which were fixed. 
I am not sure it is because of a failed kinit. I just got failed kinits but managed to send data and there were no leaks (maybe someone change something on the setup while I was working on it).
The issue happens whenever I get delivery errors (and kerberos parameters are set).
The memory usage increases constantly: 
20180813_153430: VmRSS: 46,051,448 kB (+538108) 34.8439%
20180813_153530: VmRSS: 46,659,796 kB (+608348) 35.3042%
20180813_153630: VmRSS: 47,243,380 kB (+583584) 35.7458%
20180813_153730: VmRSS: 47,826,864 kB (+583484) 36.1872%
20180813_153830: VmRSS: 48,399,992 kB (+573128) 36.6209%
20180813_153930: VmRSS: 48,912,700 kB (+512708) 37.0088%
20180813_154030: VmRSS: 49,544,460 kB (+631760) 37.4868%
20180813_154130: VmRSS: 50,117,040 kB (+572580) 37.9201%
20180813_154230: VmRSS: 50,608,876 kB (+491836) 38.2922%
20180813_154330: VmRSS: 51,169,088 kB (+560212) 38.7161%
20180813_154430: VmRSS: 51,758,328 kB (+589240) 39.1619%
20180813_154530: VmRSS: 52,334,964 kB (+576636) 39.5982%
20180813_154630: VmRSS: 53,070,180 kB (+735216) 40.1545%
20180813_154730: VmRSS: 53,668,464 kB (+598284) 40.6072%

In this specific run, there was a big free of 24GB in one minute: 
20180813_164331: VmRSS: 65,076,720 kB (+168) 49.2390%
20180813_164431: VmRSS: 65,077,072 kB (+352) 49.2393%
20180813_164531: VmRSS: 65,077,216 kB (+144) 49.2394%
20180813_164631: VmRSS: 40,090,656 kB (-24986560) 30.3338%
20180813_164731: VmRSS: 40,800,240 kB (+709584) 30.8707%
20180813_164831: VmRSS: 41,616,140 kB (+815900) 31.4880%
20180813_164931: VmRSS: 42,267,064 kB (+650924) 31.9805%
20180813_165031: VmRSS: 43,096,352 kB (+829288) 32.6080%

When running with different settings (removing the kerberos parameters), the memory usage is around 28% for days (whether I constantly get delivery fail or constantly get delivery success).

btw, additionaly to above, I have these parameters set:

"produce.offset.report" = "true"
"message.timeout.ms" = "6000"

Thanks,
Dror

**Comment by drorgu** [comment]
Ok, it looks like the issue is solved by configuring a higher batch.num.messages and queue.buffering.max.ms. 
They were set to:
{"Param":"batch.num.messages","Value":"50"},
{"Param":"queue.buffering.max.ms","Value":"200"},
so I changed them both to:
{"Param":"batch.num.messages","Value":"10000"}, --default
{"Param":"queue.buffering.max.ms","Value":"10000"}, - 10 second in case idle

It makes sense that that's what caused a performance issue... but there were some symptoms I cannot still understand: 
1. I have 7200% CPU (72 cores) on my machine and my program used 2000% max. Still other threads had a hard time and dropped data. There are no mutual mutexes shared with the main thread using librdkafka.
2. The memory consumption which kept on increasing (until program restarted itself due to RAM usage close to 100% - 132GB on my machine). If I run without librdkafka and simulate traffic partially dropping on other threads and other stress scenarios, there is no memory leak.

So I still don't completely understand this behaviour (which still might well be another issue on my side, though I have looked almost everywhere...).

Anyway, this ticket can be closed.

Thanks,
Dror
